api/src/main/java/jakarta/data/repository/Is.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2024 Contributors to the Eclipse Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package jakarta.data.repository;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * <p>Annotates a parameter of a repository {@link Find} or {@link Delete} method,
+ * indicating how a persistent field is compared against the parameter's value.
+ * The {@link By} annotation is used on the same parameter to identify the
+ * persistent field.</p>
+ *
+ * <p>For example,</p>
+ *
+ * <pre>
+ * &#64;Repository
+ * public interface Products extends CrudRepository&lt;Product, Long&gt; {
+ *
+ *     // Find all Product entities where the price field is less than a maximum value.
+ *     &#64;Find
+ *     List&lt;Product&gt; pricedBelow(&#64;By(_Product.PRICE) &#64;Is(LESS_THAN) float max);
+ *
+ *     // Find a page of Product entities where the name field matches a pattern, ignoring case.
+ *     &#64;Find
+ *     Page&lt;Product&gt; search(&#64;By(_Product.NAME) &#64;Is(LIKE_IGNORE_CASE) String pattern,
+ *                          PageRequest pagination,
+ *                          Order&lt;Product&gt; order);
+ *
+ *     // Remove Product entities with any of the unique identifiers listed.
+ *     &#64;Delete
+ *     void remove(&#64;By(ID) &#64;Is(IN) List&lt;Long&gt; productIds);
+ * }
+ * </pre>
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.PARAMETER)
+public @interface Is {
+    // TODO add JavaDoc with examples to these
+    String EQUAL                = "EQUAL";
```

Comment:
I understand the power of flexibility of `String`; however, a typo can be a disaster; plus, the enum can show the available options by default.


In this case, I still prefer the enum; if a vendor wants to externalize this behavior, it can easily create its annotation.

Reply:
I changed this to make it more agreeable to Gavin.  My own preference was for enum which I do still prefer, but I don't think it is a big deal one way or the other and would be happy with either.  If we did leave it this way, I'm not t all worried about typos because the user doesn't write these, they use the constants we define in the spec (or that a vendor defines in their API).  And yes, having a vendor extend behavior with an annotation of their seems perfectly reasonable to me, but I don't think Gavin is going to be on board with that and he was the one who brought up not being extensible as a drawback.  In any case, if the two of you can agree on one way or the other (enum or String constant), I'll switch it to that because I really am fine with either.

Reply:
I would vote to keep an enum or reuse the `By` annotation to give the provider more flexibility.


https://github.com/jakartaee/data/issues/857#issuecomment-2461965944

Reply:
That's fine, I put the enum back for now. I was only anticipating that Gavin would prefer it with String constants to enable vendor extensions without an annotation, but I'm not actually sure of that.  This is also now updated with the shorter `ANY_CASE`, and I added in the full list of possible enum constants, which I realize might need to be pared down later to gain agreement on getting this in.